File: examples/php/ws-example.php
<?php 

$root = dirname (dirname (dirname (__FILE__)));

include $root . '/ccxt.php';
include $root . '/php/async/WebsocketConnection.php';

require __DIR__ . '/../../vendor/autoload.php';

$loop = React\EventLoop\Factory::create();


$ws = new ccxt\WebsocketConnection([
  // 'url' => 'wss://ws.cex.io/ws/'
  'url' => 'wss://echo.websocket.org'
], 1000, $loop);
/*
try {
  $value = Clue\React\Block\await($ws->connect(), $loop);
  // promise successfully fulfilled with $value
  echo 'Result: ' . $value;
} catch (Exception $exception) {
  // promise rejected with $exception
  echo 'ERROR: ' . $exception->getMessage();
}
*/

$ws->on ('err', function ($error) {
    echo ($error);
});

$ws->on ('message', function ($msg) use ($ws, $loop){
    $loop->addTimer(5, function () use ($ws) {
        $ws->sendJson ([
            'helo' => 'helo'
        ]);
    });
});

$ws->on ('close', function ($msg) {
    echo '--->close';
});

try {
    Clue\React\Block\await ($ws->connect(), $loop);
    $deferred = new \React\Promise\Deferred();
    $ws->once ('message', function ($msg) use ($deferred){
        $deferred->resolve($msg);
    });
    $ws->sendJson ([
        'helo' => 'helo'
    ]);
    /*
    $promise = new React\Promise\Promise (
        function ($resolve, $reject) use($ws) {
            $ws->once ('message', function ($msg) use ($resolve) {
                // $resolve($msg);
            });
                    
        });
        */
    $timeout = React\Promise\Timer\timeout($deferred->promise(), 5.0, $loop);
        
    $value = Clue\React\Block\await($timeout, $loop);
    echo 'awaited for:' . $value;
      
} catch (Exception $ex) {
    echo $ex;
    $ws->close();
}

/*
$ws->connect()->then(function () use($loop, $ws) {
  $ws->sendJson ([
        'helo' => 'helo'
  ]);
  
  $promise = new React\Promise (
    function ($resolve, $reject) use($ws) {
        echo 'ññññ';
        $ws->on ('message', function ($msg) {
            echo 'message';
            $resolve($msg);
        });
    });
    $timeout = React\Promise\Timer\timeout($promise, 10.0, $loop);
    
    // Clue\React\Block\sleep(5, $loop);
    try {
        $value = Clue\React\Block\await($promise, $loop);
        echo $value;
    } catch (Exception $ex) {
        echo $ex;
    }
    $ws->close();
});
*/
$loop->run();

File: examples/py/ws-example.py
import asyncio
import os
import sys

root = os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
sys.path.append(root + '/python')

# import ccxt.async as ccxt  # noqa: E402
from ccxt.async_support.websocket.websocket_connection import WebsocketConnection  # noqa: E402
loop = asyncio.get_event_loop()  # type: asyncio.BaseEventLoop


async def main():
    ws = WebsocketConnection({
        'url': 'wss://echo.websocket.org'
    }, 5 * 1000, loop)

    @ws.on('err')
    def error_handler(error):  # pylint: disable=W0612
        print(error)
        loop.stop()

    @ws.on('message')
    def message_handler(msg):  # pylint: disable=W0612
        print(msg)
        sys.stdout.flush()
        # ws.close()
        # loop.stop()
        print("Waiting 2 seconds...")
        sys.stdout.flush()
        loop.call_later(2, lambda: ws.sendJson({'helo': 'helo'}))

    @ws.on('close')
    def close_handler():  # pylint: disable=W0612
        print('--->closed')
        sys.stdout.flush()
        loop.stop()

    # loop.run_until_complete(ws.connect())
    await ws.connect()
    print("after run")
    ws.sendJson({'helo': 'helo'})

loop.run_until_complete(main())
loop.run_forever()
print("after complete")
